### ISSUE TITLE
Attribute backed by descriptor is unknown when imported from external package

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#1518]
A library defines `Foo` and `Bar` like this:

```python
from typing import Any

class Foo:
    def __get__(self, *args: Any) -> str:
        return "foo"

class Bar:
    foo = Foo()
```

Accessing the `foo` attribute in my project gives an error message in strict mode:

```python
Bar().foo  # Type of "foo" is unknown
```

If I define `Foo` and `Bar` in my project instead the error goes away.

It's possible to reproduce the issue by injecting the sample code above in an arbitrary `py.typed` library:

```bash
mkdir foo
cd foo
python -m venv .venv
source .venv/bin/activate
pip install pydantic  # arbitrary py.typed library to replicate the issue into
npm i pyright
echo '{"typeCheckingMode":"strict"}' > pyrightconfig.json
```

Then create a file with the following content:

```python
# foo.py
from pydantic import Bar

Bar().foo
```

And in vscode control-click on `pydantic` to access the `__init__.py` file and paste the sample code at the end:

```python
# pydantic/__init__.py
from typing import Any

class Foo:
    def __get__(self, *args: Any) -> str:
        return "foo"

class Bar:
    foo = Foo()
```

Now running pyright reports the error:

```bash
$ npx pyright
/tmp/foo/foo.py
  3:1 - error: Type of "foo" is unknown (reportUnknownMemberType)
```

However if we define `Foo` and `Bar` in the file we created the error goes away:

```python
# foo.py
from typing import Any

class Foo:
    def __get__(self, *args: Any) -> str:
        return "foo"

class Bar:
    foo = Foo()

Bar().foo
```

```bash
$ npx pyright
0 errors, 0 warnings, 0 infos
```

It took me a while to boil it down to this. I'm not sure if this is because inside the project Pyright is able to infer something that would normally necessitate an annotation or if there's really a bug.

In my case I have dozens of these attributes, all with complicated generic types, and the descriptor packages the necessary plumbing to keep the class definition clean.

**Comment by erictraut** [comment]
This is by design. If your package claims that it is typed (i.e. contains a "py.typed" file as per PEP 561), then it needs to provide types. Pyright will not attempt to infer types within a package that claims it is typed. Type inference rules are not standardized, and different type checkers infer types in slightly different ways, so it would be very bad if a package author were to rely on type inference to define the public contract for their library.

Please refer to [this guidance](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md) for library authors. The guidelines in this documentation have been discussed and generally agreed upon by maintainers of all the major Python type checkers.

The command-line version of pyright includes an option for verifying a "py.typed" library and computing its "type completeness". It also emits errors for any symbols that are part of the public interface contract that lack type annotations.

**Comment by vberlier** [comment]
Ouch that's really inconvenient. I mean the descriptor is already typed, so it just makes things really verbose for nothing. In situations where the values passed to the descriptor need to be retrieved as-is you would end up with something really hard to maintain:

```python
class A:
    foo = Thing("param1", flag=False, args=("one", "two", foo_validator))

class A:
    foo: ClassVar[
        Thing[
            Literal["param1"],
            Literal[False],
            Tuple[Literal["one"], Literal["two"], Callable[..., bool]],
        ]
    ] = Thing("param1", flag=False, args=("one", "two", foo_validator))
```

I'm looking through the guidance and to me it seems like to achieve type completeness you need to steer away from inline types. I'm guessing the best solution to avoid littering unmaintainable annotations everywhere is to "bake" the inference rules of pyright by generating `.pyi` files and including them in the package. That's definitely not a low barrier to entry.

**Comment by erictraut** [comment]
Type annotations are almost never that verbose. The example above doesn't look correct to me, but I can't say for sure because I don't know what your intent is.

Is `Thing` in your example really a generic class that has three type parameters? If so, is this particular variable really parameterized with literal type arguments? And is `foo` really a ClassVar (i.e. meant to be accessible only through a class reference)? Those would all be really strange design choices.

Here's a more typical type annotation.

```python
class A:
    foo: Thing = Thing("param1", flag=False, args=("one", "two", foo_validator))
```

Speaking from personal experience, type annotations do not typically add much verboseness, nor are they difficult to maintain. On the contrary, they make it easier to maintain your code if you use a type checker because changes that break type contracts are immediately highlighted. Type annotations need to be added only once, and then you reap the benefits from then on.

If you're not fully sold on type annotations, keep in mind that you need to annotate only the public interface for your library, not any of the internal implementation details. For most libraries, the public interface is a small fraction of the overall code.

Out of curiosity, are you the maintainer of a library? If so, which one?

Are there any other tools, support or guidance that would help you out?

**Comment by vberlier** [comment]
You're right the example doesn't need the `ClassVar` and is a bit exaggerated. However I'm actually facing a problem right now because I'm using the descriptor in a dataclass:

```python
from dataclasses import dataclass
from typing import Any, ClassVar

class Foo:
    def __get__(self, *args: Any) -> str:
        return "foo"

    def __set__(self, obj: Any, value: str):
        print(obj, value)

@dataclass
class Bar1:
    name: str
    foo = Foo()

b = Bar1("hello")
b.foo = "foo"  # Bar1(name='hello') foo

@dataclass
class Bar2:
    name: str
    foo: Foo = Foo()

b = Bar2("hello")  # Bar2(name='hello', foo='foo') foo
b.foo = "foo"  # Bar2(name='hello', foo='foo') foo
```

So I'm in a situation where adding the annotation actually changes the constructor, the repr and everything else on top of generating an additional invalid call to the setter during the constructor.

I'm very much sold on type annotations, I find typescript invaluable but when it comes to python there are still a lot of rough edges.

**Comment by erictraut** [comment]
Wow, this is really weird. Do you understand why the introduction of that type annotation is changing the behavior? It's not obvious to me.

One option here is to fall back on a type comment:

```python
@dataclass
class Bar2:
    name: str
    foo = Foo()  # type: Foo
```

**Comment by vberlier** [comment]
I'm guessing the dataclass thinks that it's simply another field with a default value and the generated constructor somehow manages to call the setter with the value returned by the getter...? I agree there's something weird going on.

Didn't think about the type comment though, thanks!

**Comment by erictraut** [comment]
Ah, I think I understand. When you add the type annotation to it, the dataclass constructor is treating it as an instance variable and is instantiated once every time the dataclass is instantiated. Without the type annotation, it's treated as a class variable, and the descriptor is instantiated only once.

So another (probably better) option is:
```python
@dataclass
class Bar2:
    name: str
    foo: ClassVar[Foo] = Foo()
```

This exposes a bug in pyright though. I had assumed that a variable marked as ClassVar could not be accessed through a class instance, but that is apparently not the case. So the line `b.foo = "foo"` works fine at runtime, but pyright generates an error. I will fix that.

**Comment by vberlier** [comment]
Actually yeah I was trying that earlier that's why I posted an example with `ClassVar`. I wasn't sure whether the error was intended or not but it seemed to work fine at runtime.

**Comment by erictraut** [comment]
I think the error is correct for a non-descriptor attribute, but it's incorrect for a descriptor. Unfortunately, mypy also flags this as an error, so it will need to be fixed there too if you use mypy.

**Comment by vberlier** [comment]
Hmm I guess in the meantime the best solution is the type comment then. Right?

**Comment by erictraut** [comment]
Yeah, that's the best workaround I can think of for now.

**Comment by vberlier** [comment]
Awesome! Thank you so much for you help :)

**Comment by erictraut** [comment]
I've modified the logic for enforcing ClassVar attributes. A write to the attribute is now allowed through a class instance if it's through a descriptor. It's still disallowed in other cases.

```python
@dataclass
class Bar2:
    name: str
    foo: ClassVar[Foo] = Foo()
    bar: ClassVar[str] = "hi"


b2 = Bar2("hello")
b2.foo = "" # No Error
b2.bar = "" # Error: Member "bar" cannot be assigned through a class instance because it is a ClassVar
```

This fix will be included in the next release of pyright.

**Comment by vberlier** [comment]
Sounds great! Does this mean all descriptors should be annotated with `ClassVar` or only when they conflict with something else like dataclasses?

**Comment by erictraut** [comment]
In contexts other than dataclass, it should work fine with or without `ClassVar`, so it's up to you.

**Comment by vberlier** [comment]
I was kind of hoping that there would be some sort of convention, I'd say `ClassVar` technically makes sense because the descriptor is bound to the class and not the instance. But it adds unnecessary noise to the annotation so I don't know. Anyway, thanks again!

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.114, which I just published. It will also be in the next release of Pylance.
